### PR TITLE
nixos/matrix-sliding-sync: init

### PR DESCRIPTION
**Pull request by SuperSandro2000**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/matrix/matrix-sliding-sync.nix [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    port = lib.mkOption {
+      type = lib.types.port;
+      default = 8449;
+      description = lib.mdDoc ''
+        The port under which the sliding sync proxy should listen.
+      '';
+    };
+
+    environmentFile = lib.mkOption {
```
Comment (emilylange) [inline nixos/modules/services/matrix/matrix-sliding-sync.nix:26]: I would prefer a dedicated secret file specifically for `SYNCV3_SECRET` instead.
Or even dropping this altogether in favor of creating that secret dynamically on first start.

I might do a follow-up PR for that some time in the future

**Review by emilylange — CHANGES_REQUESTED** [review]
How about the following instead?

```patch
diff --git a/nixos/modules/services/matrix/matrix-sliding-sync.nix b/nixos/modules/services/matrix/matrix-sliding-sync.nix
index c5776c27051..73e685f99a8 100644
--- a/nixos/modules/services/matrix/matrix-sliding-sync.nix
+++ b/nixos/modules/services/matrix/matrix-sliding-sync.nix
@@ -5,58 +5,88 @@ let
 in
 {
   options.services.matrix-synapse.sliding-sync = {
-    enable = lib.mkEnableOption (lib.mdDoc "Sliding Sync Proxy");
+    enable = lib.mkEnableOption (lib.mdDoc "sliding sync");
 
     package = lib.mkPackageOption pkgs "matrix-sliding-sync" { };
 
-    listenAddress = lib.mkOption {
-      type = lib.types.str;
-      default = "127.0.0.1";
-      description = "The address sliding sync proxy should listen.";
-    };
-
-    port = lib.mkOption {
-      type = lib.types.port;
-      default = 8449;
-      description = lib.mdDoc ''
-        The port under which the sliding sync proxy should listen.
+    settings = lib.mkOption {
+      description = ''
+        Free-form environment variables passed to the sliding sync proxy.
+        Refer to <https://github.com/matrix-org/sliding-sync#setup> for
+        all supported values.
       '';
+      type = lib.types.submodule {
+        freeformType = with lib.types; attrsOf str;
+        options = {
+          SYNCV3_SERVER = lib.mkOption {
+            type = lib.types.str;
+            description = lib.mdDoc ''
+              The destination homeserver to talk to (CS API HTTPS URL)
+              e.g `https://matrix-client.matrix.org` (not including `/_matrix/`).
+            '';
+          };
+
+          SYNCV3_DB = lib.mkOption {
+            type = lib.types.str;
+            default = "postgresql:///matrix-sliding-sync?host=/run/postgresql";
+            description = lib.mdDoc ''
+              The postgres connection string.
+              Refer to <https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING>.
+            '';
+          };
+
+          SYNCV3_BINDADDR = lib.mkOption {
+            type = lib.types.str;
+            default = "127.0.0.1:8009";
+            example = "[::]:8008";
+            description = lib.mdDoc "The interface and port to listen on.";
+          };
+
+          SYNCV3_LOG_LEVEL = lib.mkOption {
+            type = lib.types.enum [ "trace" "debug" "info" "warn" "error" "fatal" ];
+            default = "info";
+            description = lib.mdDoc "The level of verbosity for messages logged.";
+          };
+        };
+      };
+      default = { };
     };
 
+    createDatabase = lib.mkEnableOption
+      (lib.mdDoc "{option}`services.postgres` and create the necessary database")
+      // { default = true; };
+
     environmentFile = lib.mkOption {
       type = lib.types.str;
       description = lib.mdDoc ''
         Environment file as defined in {manpage}`systemd.exec(5)`.
 
-        This must contain the `SYNCV3_SECRET` variable which should
-        be generated with ``openssl rand -hex 32``.
+        This must contain the {env}`SYNCV3_SECRET` variable which should
+        be generated with {command}`openssl rand -hex 32`.
       '';
     };
   };
 
   config = lib.mkIf cfg.enable {
-    services.postgresql = {
+    services.postgresql = lib.optionalAttrs cfg.createDatabase {
+      enable = lib.mkDefault true;
       ensureDatabases = [ "matrix-sliding-sync" ];
-      ensureUsers = [ rec {
+      ensureUsers = [rec {
         name = "matrix-sliding-sync";
         ensurePermissions."DATABASE \"${name}\"" = "ALL PRIVILEGES";
-      } ];
+      }];
     };
 
     systemd.services.matrix-sliding-sync = {
-      after = [ "matrix-synapse.service" "postgresql.service" ];
+      after = lib.optional cfg.createDatabase "postgresql.service";
       wantedBy = [ "multi-user.target" ];
+      environment = cfg.settings;
       serviceConfig = {
         DynamicUser = true;
         StateDirectory = "matrix-sliding-sync";
-        WorkingDirectory = "/var/lib/matrix-sliding-sync";
-        Environment = [
-          "SYNCV3_SERVER=${config.services.matrix-synapse.settings.public_baseurl}"
-          "SYNCV3_DB=postgresql:///matrix-sliding-sync?host=/run/postgresql"
-          "SYNCV3_BINDADDR=${cfg.listenAddress}:${toString cfg.port}"
-        ];
+        WorkingDirectory = "%S/matrix-sliding-sync";
         EnvironmentFile = cfg.environmentFile;
-        ExecStart = "${cfg.package}/bin/syncv3";
+        ExecStart = lib.getExe cfg.package;
       };
     };
   };
```

File: nixos/modules/services/matrix/matrix-sliding-sync.nix (hunk 1 above)
Comment (n0emis) [reply nixos/modules/services/matrix/matrix-sliding-sync.nix:26]: hmm, i would prefer to at least have a built-in way to specify a secret file; for migrating etc. I prefer to always have all secrets deployed, so there is no state (except database) on the systems which is needed for running

**Comment by SuperSandro2000** [comment]
Works after manually pointing element to the service domain. Not sure what's wrong here, probably something off with my .well-known files.

File: nixos/modules/services/matrix/matrix-sliding-sync.nix [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    services.postgresql = {
+      ensureDatabases = [ "matrix-sliding-sync" ];
+      ensureUsers = [ rec {
+        name = "matrix-sliding-sync";
+        ensurePermissions."DATABASE \"${name}\"" = "ALL PRIVILEGES";
+      } ];
+    };
+
+    systemd.services.matrix-sliding-sync = {
+      after = [ "matrix-synapse.service" "postgresql.service" ];
```
[the remark this answers is not in the document]
Comment (emilylange) [reply nixos/modules/services/matrix/matrix-sliding-sync.nix]: Technically speaking, this sliding sync proxy isn't limited to synapse, but instead could also be used with `services.dendrite` and `services.conduit`. Though the latter is currently in the progress implementing sliding sync natively.

File: nixos/modules/services/matrix/matrix-sliding-sync.nix [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        Freeform environment variables passed to the sliding sync proxy.
+        Refer to <https://github.com/matrix-org/sliding-sync#setup> for all supported values.
+      '';
+    };
+
+    createDatabase = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      description = lib.mdDoc ''
+        Wether to enable and configure `services.postgres` to ensure that the database user `matrix-sliding-sync`
```
Comment (emilylange) [inline nixos/modules/services/matrix/matrix-sliding-sync.nix:57]: typo
```suggestion
        Whether to enable and configure `services.postgres` to ensure that the database user `matrix-sliding-sync`
```

**Comment by emilylange** [comment]
> Works after manually pointing element to the service domain. Not sure what's wrong here, probably something off with my .well-known files.

Works for me™

But jokes aside, your `.well-known` looks good to me and the sliding sync proxy responding properly on the necessary paths as well.
Maybe your client being a bit weird then?

**Comment by SuperSandro2000** [comment]
> Maybe your client being a bit weird then?

or cors/csp or something

**Review by emilylange — APPROVED** [review]

**Pull request merged**